File: Java/labs/labs12_Flyweight/task_2_2/src/Engine.java
public class Engine {
    public enum Fuel {
        Petrol,
        Diesel,
        Electric
    }

    final int power;
    final Fuel fuel;

    public Engine(int power, Fuel fuel) {
        this.power = power;
        this.fuel = fuel;
    }

    @Override
    public String toString() {
        return "Engine{" +
                "power=" + power +
                ", fuel=" + fuel +
                '}';
    }
}

File: Cpp/lab12_Flyweight/task_2_2/src/engine.cpp
#include "engine.hpp"


Engine::Engine(int power, Engine::Fuel const& fuel)
: power(power), fuel(fuel) {}

Engine::~Engine() {}

std::ostream& operator<<(std::ostream& os, Engine const& engine)
{
  os << "Engine{power=" << engine.power << ", fuel=";

  switch (engine.fuel)
  {
    case Engine::Fuel::Petrol:
      os << "Petrol";
      break;
    case Engine::Fuel::Diesel:
      os << "Diesel";
      break;
    case Engine::Fuel::Electric:
      os << "Electric";
      break;
  }

  os << "}";

  return os;
}
